File: examples/c/cloud-simple/cloud-simple.c
/* Copyright (c) 2007-2022. The SimGrid Team. All rights reserved.          */

/* This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. */

#include "simgrid/actor.h"
#include "simgrid/engine.h"
#include "simgrid/host.h"
#include "simgrid/mailbox.h"
#include "simgrid/plugins/live_migration.h"
#include "simgrid/vm.h"
#include "xbt/log.h"
#include "xbt/str.h"
#include "xbt/sysdep.h"

XBT_LOG_NEW_DEFAULT_CATEGORY(cloud_simple, "Messages specific for this example");

static void computation_fun(int argc, char* argv[])
{
  const char* pr_name   = sg_actor_get_name(sg_actor_self());
  const char* host_name = sg_host_get_name(sg_host_self());

  double clock_sta = simgrid_get_clock();
  sg_actor_execute(1000000);
  double clock_end = simgrid_get_clock();

  XBT_INFO("%s:%s task executed %g", host_name, pr_name, clock_end - clock_sta);
}

static void launch_computation_worker(sg_host_t host)
{
  sg_actor_create("compute", host, computation_fun, 0, NULL);
}

struct task_priv {
  sg_host_t tx_host;
  sg_actor_t tx_proc;
  double clock_sta;
};

static void communication_tx_fun(int argc, char* argv[])
{
  xbt_assert(argc == 2);
  sg_mailbox_t mbox = sg_mailbox_by_name(argv[1]);

  struct task_priv* priv = xbt_new(struct task_priv, 1);
  priv->tx_proc          = sg_actor_self();
  priv->tx_host          = sg_host_self();
  priv->clock_sta        = simgrid_get_clock();

  sg_mailbox_put(mbox, priv, 1000000);
}

static void communication_rx_fun(int argc, char* argv[])
{
  const char* pr_name   = sg_actor_get_name(sg_actor_self());
  const char* host_name = sg_host_get_name(sg_host_self());
  xbt_assert(argc == 2);
  sg_mailbox_t mbox = sg_mailbox_by_name(argv[1]);

  struct task_priv* priv = (struct task_priv*)sg_mailbox_get(mbox);
  double clock_end       = simgrid_get_clock();

  XBT_INFO("%s:%s to %s:%s => %g sec", sg_host_get_name(priv->tx_host), sg_actor_get_name(priv->tx_proc), host_name,
           pr_name, clock_end - priv->clock_sta);

  free(priv);
}

static void launch_communication_worker(sg_host_t tx_host, sg_host_t rx_host)
{
  char* mbox = bprintf("MBOX:%s-%s", sg_host_get_name(tx_host), sg_host_get_name(rx_host));

  const char* tx_argv[] = {"comm_tx", mbox, NULL};
  sg_actor_create_("comm_tx", tx_host, communication_tx_fun, 2, tx_argv);

  const char* rx_argv[] = {"comm_rx", mbox, NULL};
  sg_actor_create_("comm_rx", rx_host, communication_rx_fun, 2, rx_argv);

  xbt_free(mbox);
}

static void master_main(int argc, char* argv[])
{
  sg_host_t pm0 = sg_host_by_name("Fafard");
  sg_host_t pm1 = sg_host_by_name("Tremblay");
  sg_host_t pm2 = sg_host_by_name("Bourassa");

  XBT_INFO("## Test 1 (started): check computation on normal PMs");

  XBT_INFO("### Put a task on a PM");
  launch_computation_worker(pm0);
  sg_actor_sleep_for(2);

  XBT_INFO("### Put two tasks on a PM");
  launch_computation_worker(pm0);
  launch_computation_worker(pm0);
  sg_actor_sleep_for(2);

  XBT_INFO("### Put a task on each PM");
  launch_computation_worker(pm0);
  launch_computation_worker(pm1);
  sg_actor_sleep_for(2);

  XBT_INFO("## Test 1 (ended)");

  XBT_INFO("## Test 2 (started): check impact of running a task inside a VM (there is no degradation for the moment)");

  XBT_INFO("### Put a VM on a PM, and put a task to the VM");
  sg_vm_t vm0 = sg_vm_create_core(pm0, "VM0");
  sg_vm_start(vm0);
  launch_computation_worker((sg_host_t)vm0);
  sg_actor_sleep_for(2);
  sg_vm_destroy(vm0);

  XBT_INFO("## Test 2 (ended)");

  XBT_INFO(
      "## Test 3 (started): check impact of running a task collocated with a VM (there is no VM noise for the moment)");

  XBT_INFO("### Put a VM on a PM, and put a task to the PM");
  vm0 = sg_vm_create_core(pm0, "VM0");
  sg_vm_start(vm0);
  launch_computation_worker(pm0);
  sg_actor_sleep_for(2);
  sg_vm_destroy(vm0);

  XBT_INFO("## Test 3 (ended)");

  XBT_INFO("## Test 4 (started): compare the cost of running two tasks inside two different VMs collocated or not (for"
           " the moment, there is no degradation for the VMs. Hence, the time should be equals to the time of test 1");

  XBT_INFO("### Put two VMs on a PM, and put a task to each VM");
  vm0         = sg_vm_create_core(pm0, "VM0");
  sg_vm_t vm1 = sg_vm_create_core(pm0, "VM1");
  sg_vm_start(vm0);
  sg_vm_start(vm1);
  launch_computation_worker((sg_host_t)vm0);
  launch_computation_worker((sg_host_t)vm1);
  sg_actor_sleep_for(2);
  sg_vm_destroy(vm0);
  sg_vm_destroy(vm1);

  XBT_INFO("### Put a VM on each PM, and put a task to each VM");
  vm0 = sg_vm_create_core(pm0, "VM0");
  vm1 = sg_vm_create_core(pm1, "VM1");
  sg_vm_start(vm0);
  sg_vm_start(vm1);
  launch_computation_worker((sg_host_t)vm0);
  launch_computation_worker((sg_host_t)vm1);
  sg_actor_sleep_for(2);
  sg_vm_destroy(vm0);
  sg_vm_destroy(vm1);
  XBT_INFO("## Test 4 (ended)");

  XBT_INFO("## Test 5  (started): Analyse network impact");
  XBT_INFO("### Make a connection between PM0 and PM1");
  launch_communication_worker(pm0, pm1);
  sg_actor_sleep_for(5);

  XBT_INFO("### Make two connection between PM0 and PM1");
  launch_communication_worker(pm0, pm1);
  launch_communication_worker(pm0, pm1);
  sg_actor_sleep_for(5);

  XBT_INFO("### Make a connection between PM0 and VM0@PM0");
  vm0 = sg_vm_create_core(pm0, "VM0");
  sg_vm_start(vm0);
  launch_communication_worker(pm0, (sg_host_t)vm0);
  sg_actor_sleep_for(5);
  sg_vm_destroy(vm0);

  XBT_INFO("### Make a connection between PM0 and VM0@PM1");
  vm0 = sg_vm_create_core(pm1, "VM0");
  sg_vm_start(vm0);
  launch_communication_worker(pm0, (sg_host_t)vm0);
  sg_actor_sleep_for(5);
  sg_vm_destroy(vm0);

  XBT_INFO("### Make two connections between PM0 and VM0@PM1");
  vm0 = sg_vm_create_core(pm1, "VM0");
  sg_vm_start(vm0);
  launch_communication_worker(pm0, (sg_host_t)vm0);
  launch_communication_worker(pm0, (sg_host_t)vm0);
  sg_actor_sleep_for(5);
  sg_vm_destroy(vm0);

  XBT_INFO("### Make a connection between PM0 and VM0@PM1, and also make a connection between PM0 and PM1");
  vm0 = sg_vm_create_core(pm1, "VM0");
  sg_vm_start(vm0);
  launch_communication_worker(pm0, (sg_host_t)vm0);
  launch_communication_worker(pm0, pm1);
  sg_actor_sleep_for(5);
  sg_vm_destroy(vm0);

  XBT_INFO("### Make a connection between VM0@PM0 and PM1@PM1, and also make a connection between VM0@PM0 and VM1@PM1");
  vm0 = sg_vm_create_core(pm0, "VM0");
  vm1 = sg_vm_create_core(pm1, "VM1");
  sg_vm_start(vm0);
  sg_vm_start(vm1);
  launch_communication_worker((sg_host_t)vm0, (sg_host_t)vm1);
  launch_communication_worker((sg_host_t)vm0, (sg_host_t)vm1);
  sg_actor_sleep_for(5);
  sg_vm_destroy(vm0);
  sg_vm_destroy(vm1);

  XBT_INFO("## Test 5 (ended)");

  XBT_INFO("## Test 6 (started): Check migration impact (not yet implemented neither on the CPU resource nor on the"
           " network one");
  XBT_INFO("### Relocate VM0 between PM0 and PM1");
  vm0 = sg_vm_create_core(pm0, "VM0");
  sg_vm_set_ramsize(vm0, 1L * 1024 * 1024 * 1024); // 1GiB

  sg_vm_start(vm0);
  launch_communication_worker((sg_host_t)vm0, pm2);
  sg_actor_sleep_for(0.01);
  sg_vm_migrate(vm0, pm1);
  sg_actor_sleep_for(0.01);
  sg_vm_migrate(vm0, pm0);
  sg_actor_sleep_for(5);
  sg_vm_destroy(vm0);
  XBT_INFO("## Test 6 (ended)");
}

int main(int argc, char* argv[])
{
  /* Get the arguments */
  simgrid_init(&argc, argv);
  sg_vm_live_migration_plugin_init();

  /* load the platform file */
  simgrid_load_platform(argv[1]);

  sg_actor_create("master_", sg_host_by_name("Fafard"), master_main, 0, NULL);

  simgrid_run();
  XBT_INFO("Simulation time %g", simgrid_get_clock());

  return 0;
}

File: examples/java/cloud_simple/cloud_simple.java
/* Copyright (c) 2007-2025. The SimGrid Team. All rights reserved.          */

/* This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. */

import org.simgrid.s4u.*;

class ComputationWorker extends Actor {
  public void run()
  {
    double clockStart = Engine.get_clock();
    this.execute(1000000);
    double clockEnd = Engine.get_clock();

    Engine.info("%s:%s executed %g", this.get_host().get_name(), this.get_name(), clockEnd - clockStart);
  }
}

class Payload {
  public final Host txHost;
  public final String txActorName;
  public final double clockStart;
  Payload(Host h, String n)
  {
    txHost      = h;
    txActorName = n;
    clockStart  = Engine.get_clock();
  }
}

class CommunicationTX extends Actor {
  String mboxName;
  CommunicationTX(String mboxName) { this.mboxName = mboxName; }
  public void run()
  {
    Mailbox mbox          = this.get_engine().mailbox_by_name(mboxName);
    Payload payload       = new Payload(this.get_host(), Actor.self().get_name());

    mbox.put(payload, 1000000);
  }
}

class CommunicationRX extends Actor {
  String mboxName;
  CommunicationRX(String mboxName) { this.mboxName = mboxName; }
  public void run() throws SimgridException
  {
    String actor_name = Actor.self().get_name();
    String host_name  = this.get_host().get_name();
    Mailbox mbox      = this.get_engine().mailbox_by_name(mboxName);

    Payload payload  = (Payload)mbox.get();
    double clock_end = Engine.get_clock();

    Engine.info("%s:%s to %s:%s => %g sec", payload.txHost.get_name(), payload.txActorName, host_name, actor_name,
                clock_end - payload.clockStart);
  }
}

class Main extends Actor {
  void launch_communication_worker(Host tx_host, Host rx_host)
  {
    String mbox_name = "MBOX:" + tx_host.get_name() + "-" + rx_host.get_name();

    tx_host.add_actor("comm_tx", new CommunicationTX(mbox_name));
    rx_host.add_actor("comm_rx", new CommunicationRX(mbox_name));
  }
  public void run()
  {
    Engine e = this.get_engine();
    Host pm0 = e.host_by_name("Fafard");
    Host pm1 = e.host_by_name("Tremblay");
    Host pm2 = e.host_by_name("Bourassa");

    Engine.info("## Test 1 (started): check computation on normal PMs");

    Engine.info("### Put an activity on a PM");
    pm0.add_actor("compute", new ComputationWorker());
    this.sleep_for(2);

    Engine.info("### Put two activities on a PM");
    pm0.add_actor("compute", new ComputationWorker());
    pm0.add_actor("compute", new ComputationWorker());
    this.sleep_for(2);

    Engine.info("### Put an activity on each PM");
    pm0.add_actor("compute", new ComputationWorker());
    pm1.add_actor("compute", new ComputationWorker());
    this.sleep_for(2);

    Engine.info("## Test 1 (ended)");

    Engine.info("## Test 2 (started): check impact of running an activity inside a VM (there is no degradation for "
                + "the moment)");

    Engine.info("### Put a VM on a PM, and put an activity to the VM");
    VirtualMachine vm0 = pm0.create_vm("VM0", 1);
    vm0.start();
    vm0.add_actor("compute", new ComputationWorker());
    this.sleep_for(2);
    vm0.destroy();

    Engine.info("## Test 2 (ended)");

    Engine.info(
        "## Test 3 (started): check impact of running an activity collocated with a VM (there is no VM noise for "
        + "the moment)");

    Engine.info("### Put a VM on a PM, and put an activity to the PM");
    vm0 = pm0.create_vm("VM0", 1);
    vm0.start();
    pm0.add_actor("compute", new ComputationWorker());
    this.sleep_for(2);
    vm0.destroy();
    Engine.info("## Test 3 (ended)");

    Engine.info(
        "## Test 4 (started): compare the cost of running two activities inside two different VMs collocated or not "
        +
        "(for the moment, there is no degradation for the VMs. Hence, the time should be equals to the time of test 1");

    Engine.info("### Put two VMs on a PM, and put an activity to each VM");
    vm0 = pm0.create_vm("VM0", 1);
    vm0.start();
    VirtualMachine vm1 = pm0.create_vm("VM1", 1);
    vm0.add_actor("compute", new ComputationWorker());
    vm1.add_actor("compute", new ComputationWorker());
    this.sleep_for(2);
    vm0.destroy();
    vm1.destroy();

    Engine.info("### Put a VM on each PM, and put an activity to each VM");
    vm0 = pm0.create_vm("VM0", 1);
    vm1 = pm1.create_vm("VM1", 1);
    vm0.start();
    vm1.start();
    vm0.add_actor("compute", new ComputationWorker());
    vm1.add_actor("compute", new ComputationWorker());
    this.sleep_for(2);
    vm0.destroy();
    vm1.destroy();
    Engine.info("## Test 4 (ended)");

    Engine.info("## Test 5  (started): Analyse network impact");
    Engine.info("### Make a connection between PM0 and PM1");
    launch_communication_worker(pm0, pm1);
    this.sleep_for(5);

    Engine.info("### Make two connection between PM0 and PM1");
    launch_communication_worker(pm0, pm1);
    launch_communication_worker(pm0, pm1);
    this.sleep_for(5);

    Engine.info("### Make a connection between PM0 and VM0@PM0");
    vm0 = pm0.create_vm("VM0", 1);
    vm0.start();
    launch_communication_worker(pm0, vm0);
    this.sleep_for(5);
    vm0.destroy();

    Engine.info("### Make a connection between PM0 and VM0@PM1");
    vm0 = pm1.create_vm("VM0", 1);
    launch_communication_worker(pm0, vm0);
    this.sleep_for(5);
    vm0.destroy();

    Engine.info("### Make two connections between PM0 and VM0@PM1");
    vm0 = pm1.create_vm("VM0", 1);
    vm0.start();
    launch_communication_worker(pm0, vm0);
    launch_communication_worker(pm0, vm0);
    this.sleep_for(5);
    vm0.destroy();

    Engine.info("### Make a connection between PM0 and VM0@PM1, and also make a connection between PM0 and PM1");
    vm0 = pm1.create_vm("VM0", 1);
    vm0.start();
    launch_communication_worker(pm0, vm0);
    launch_communication_worker(pm0, pm1);
    this.sleep_for(5);
    vm0.destroy();

    Engine.info(
        "### Make a connection between VM0@PM0 and PM1@PM1, and also make a connection between VM0@PM0 and VM1@PM1");
    vm0 = pm0.create_vm("VM0", 1);
    vm1 = pm1.create_vm("VM1", 1);
    vm0.start();
    vm1.start();
    launch_communication_worker(vm0, vm1);
    launch_communication_worker(vm0, vm1);
    this.sleep_for(5);
    vm0.destroy();
    vm1.destroy();

    Engine.info("## Test 5 (ended)");
    Engine.info(
        "## Test 6 (started): Check migration impact (not yet implemented neither on the CPU resource nor on the"
        + " network one)");
    Engine.info("### Relocate VM0 between PM0 and PM1");
    vm0 = pm0.create_vm("VM0", 1);
    vm0.set_ramsize(1024 * 1024 * 1024).start(); // 1GiB

    launch_communication_worker(vm0, pm2);
    this.sleep_for(0.01);
    vm0.migrate(pm1);
    this.sleep_for(0.01);
    vm0.migrate(pm0);
    this.sleep_for(5);
    vm0.destroy();
    Engine.info("## Test 6 (ended)");
  }
}
public class cloud_simple {
  public static void main(String[] args)
  {
    Engine e = new Engine(args);
    e.plugin_vm_live_migration_init();
    e.load_platform(args[0]);

    e.host_by_name("Fafard").add_actor("main", new Main());

    e.run();

    Engine.info("Simulation ends.");

    // The following call is useless in your code, but our continuous integration uses it to track memleaks
    e.force_garbage_collection();
  }
}
